### PR TITLE
fix(detection): scan credentials beside lockfile checksums

### PR DESCRIPTION
**Pull request by JeongJaeSoon**
## Summary

- neutralize only recognized checksum/hash fields in supported lockfiles
- keep URL, module-path, and all other same-line text visible to gitleaks
- keep checksum-shaped text inside quoted values/comments byte-for-byte
- require one complete package-lock JSON snapshot before neutralizing integrity
- add same-line credential regressions for `uv.lock` scan-path and `go.sum` scan-staged/Write paths

## Reproduction

Put a credential and a valid checksum on the same inline `uv.lock` record:

```text
sdist = { url = "https://example.invalid/pkg?api_token=<credential>", hash = "sha256:<64 hex>", size = 1 }
```

On merged #145, `scan-path uv.lock` exits 0 because `filter_lockfile_records()` recognizes the checksum record and drops the entire line. The same credential in a normal file exits 1.

The same bypass applies when a credential-bearing module path and valid `h1:` checksum share a `go.sum` line passed through staged-diff or proposed-Write scanning.

A second boundary case is quoted non-field text such as:

```text
note = 'api_token = { checksum = "<64 hex>" }'
```

A comma or brace inside that string was previously treated as structural, so the nested text was neutralized even though it was not a Cargo checksum field.

The same context loss occurs for staged, `apply_patch`, and Edit/MultiEdit TOML
fragments when the unchanged opening triple quote is outside the fragment. A
supported custom gitleaks rule can detect the raw 64-hex credential but receives
`CHECKSUM` after context-free structural parsing. The bundled default rules do
not independently flag this exact nested sample; the regression covers the
preprocessor boundary for supported custom rules.

## Root cause

The path-specific allowlist correctly recognizes checksum grammar, but its action is `next`, which removes every byte of the recognized record before gitleaks sees it. The first field-only replacement then matched structural delimiters with regular expressions alone, without tracking whether they occurred inside a quoted value or comment.

For complete Cargo content, treating every `checksum` assignment as a package
checksum also exceeded the lockfile schema. The generated checksum field belongs
to `[[package]]`, legacy `[root]`, or `[[patch.unused]]`; a custom table such as
`[credentials]` must remain scannable. Likewise, uv only generates these hashes
inside URL-bearing `sdist` and `wheels` artifact objects, not arbitrary inline
tables. Package-lock integrity fields likewise belong under its top-level
`packages` or `dependencies` maps rather than arbitrary root objects.

The streaming package-lock parser also emitted replacements before it knew that
the complete document was valid. Validating and filtering by reopening the same
caller-owned path would still leave a check/use race: the path could be replaced
with truncated content after validation but before filtering.

## Change

- replace only the matched checksum/hash field with a fixed `CHECKSUM` token
- for `go.sum`, anchor replacement to the third field so an earlier h1-shaped module/path token remains scannable
- require exact recognized field boundaries; `api_checksum`, `api_hash`, and `api_integrity` remain byte-for-byte scannable
- reject whitespace-separated lookalikes such as `api checksum` and require a delimiter/end after the checksum value
- track JSON/TOML/Yarn quote, escape, comment, and TOML triple-quote state so only fields at real structural positions are replaced
- leave staged/apply_patch Cargo and uv TOML fragments fully scannable because unchanged multiline-string context is unavailable; retain physical-line filtering for JSON, Yarn, and `go.sum`
- leave every Edit/MultiEdit lockfile fragment fully scannable, including `go.sum`, while preserving full-file filtering for Write
- neutralize Cargo checksums only within generated dependency tables (`[[package]]`, legacy `[root]`, and `[[patch.unused]]`)
- neutralize uv hashes only within depth-one, URL-bearing artifact objects on generated `sdist`/`wheels` records under `[[package]]` (or the legacy `[[distribution]]` alias), including the generated multi-line wheels array
- snapshot package-lock input once, validate that immutable snapshot as exactly one complete top-level JSON object, then neutralize integrity only as a direct package-entry field inside the generated top-level `packages` map or recursive `dependencies` entry maps
- keep malformed, truncated, multi-document, minified, oversized, and fragmentary JSON fully scannable; missing `jq` also falls back to unchanged content
- if private snapshot copying fails, stream the original package-lock bytes raw instead of letting an empty filter pipeline appear clean
- parse records up to 131,072 characters in one pass with 256-character field lookahead, a 256-level JSON depth bound, and 4 KiB output chunks; larger/deeper/invalid records remain unchanged and fully scannable
- replace every recognized checksum field on a compact record, not only the first
- retain the surrounding record unchanged for the downstream scan
- keep path scoping and existing checksum grammars for `go.sum`, `package-lock.json`, `yarn.lock`, `Cargo.lock`, and `uv.lock`

## Related work

- Follow-up to merged #145
- #150, #153, and #156 are separate display-redaction behavior and do not address lockfile scanning

## Functional verification

- `make test` — 590 passed, 0 failed
  - checksum-only records remain clean
  - prefixed checksum-like fields remain unchanged and detectable shapes still block
  - `api checksum: <64 hex>` remains a finding in `yarn.lock`
  - checksum-length prefixes with a trailing value byte remain unchanged
  - an h1-shaped module token remains a finding while only the harmless third-field `go.sum` checksum is neutralized
  - both checksum fields on one compact `uv.lock` record are neutralized while both URLs remain
  - checksum-shaped text inside quoted Cargo, uv, and Yarn values is preserved byte-for-byte
  - TOML multiline basic/literal strings remain unchanged across physical records while later real checksum fields are still neutralized
  - odd/even backslash parity keeps quoted text distinct from later structural fields
  - inline `uv.lock` URL credential blocks `scan-path`
  - same-line `go.sum` module-path credential blocks `scan-staged`
  - the same proposed `go.sum` Write blocks `hook-pre-tool`
  - staged, Edit/MultiEdit, and apply_patch TOML fragments preserve a nested 64-hex credential for a supported custom scanner rule instead of replacing it with `CHECKSUM`
  - a `go.sum` Edit fragment does not infer absent earlier fields and hide an h1-shaped supported custom finding
  - an `uv.lock` Edit fragment does not infer absent multiline context and hide a hash-shaped supported custom finding
  - full Write, scan-path, and untracked scans preserve a checksum-shaped value in a custom Cargo table for supported custom rules
  - full Write and scan-path scans preserve a hash-shaped value in a non-artifact uv inline table
  - direct schema probes replace Cargo checksums only in generated dependency tables and preserve a custom `[credentials]` table
  - direct schema probes replace uv hashes in compact and multi-line `sdist`/`wheels` artifact objects under `[[package]]`/`[[distribution]]`, while preserving a forged `sdist` in `[credentials]`
  - full Write and scan-path scans preserve an integrity-shaped value in a nested arbitrary package-lock object
  - staged and apply_patch package-lock fragments remain raw instead of guessing missing JSON depth
  - direct schema probes replace package-lock integrity only on package entries (including recursive v1 dependencies) while preserving nested and root arbitrary objects
  - malformed/truncated package-lock content remains unchanged and blocks in direct filtering, proposed Write, scan-path, and untracked scans
  - replacing the original package-lock path with truncated JSON immediately after validation cannot change the immutable bytes subsequently filtered
  - a forced snapshot-copy failure under closed-mode proposed Write falls back to raw scanning and blocks rather than producing clean empty stdin
  - JSON nesting beyond 256 levels switches the rest of the file to unchanged conservative scanning without dropping lines
- direct `filter-lockfile-hashes` probes preserved inline URLs, `api_token`, and prefixed user fields while changing every exact `hash = "sha256:..."` field to `hash = "sha256:CHECKSUM"`
- adversarial exact quoted-candidate probes: 3,000 candidates / 243 KiB and 10,000 candidates / 810 KiB each reported `real 0.06`; both exceed the 131,072-character filter bound and remain unchanged for conservative downstream scanning
- bounded package-lock probe: 20,000 entries / 2,808,913 bytes completed twice in `real 2.46` and `real 2.50`, including immutable snapshot, streamed whole-document validation, and field filtering
- `make smoke-test` — passed
- `scripts/validate-plugin-layout.sh` — passed
- `sh -n plugins/agent-guard/bin/agent-guard`
- `dash -n plugins/agent-guard/bin/agent-guard`
- `sh -n tests/run.sh`
- `dash -n tests/run.sh`
- `git diff --check`
- staged `agent-guard scan-staged` — passed
- `make submission-check` — expected nonzero only because the reviewed marketplace `.source.sha` does not match this unmerged plugin payload; this security PR intentionally does not pre-pin its own unmerged commit

The Makefile has no `fmt` target.

## Summary by CodeRabbit

* **Bug Fixes**
  * Made lockfile scanning more conservative: neutralizes only fully validated checksum/hash fields while preserving surrounding content.
  * Prevents rewriting when lockfile data is malformed, truncated, multi-document, or context is missing.
  * Added safer handling for NUL bytes and invalid UTF-8 during lockfile processing.
  * Improved credential scanning for edit and notebook-style payloads, independent of file paths.
* **Tests**
  * Expanded regression coverage across lockfile formats and many checksum/hash edge cases, including binary-safe inputs and malformed package-lock scenarios.

### CODE REVIEW

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 1]
```diff
@@ -817,17 +817,70 @@ run_gitleaks_stdin() {
...
+        # at a structural comma/brace, never at arbitrary whitespace that
+        # follows an attacker-controlled key phrase.
+        line = suffix_text substr(line, full_start + full_length)
+        first = 0
+      }
+      return result line
+    }
+    BEGIN {
+      field_prefix = "(^[[:space:]]*|[,{][[:space:]]*)"
+      structural_prefix = "[,{][[:space:]]*"
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Addressed in signed commit 3034cfd. The filter now replaces only fields encountered at parsed structural positions, carries TOML triple-quoted state across physical records, preserves quoted/comment text, and uses bounded 256-character lookahead plus a 131,072-character fail-conservative record cap. Cargo/uv/Yarn quote, escape-parity, multiline, later-real-field, and 3k/10k dense-candidate regressions/probes are covered; make test passes 565/0.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 2]
```diff
@@ -817,17 +817,222 @@ run_gitleaks_stdin() {
...
+        if (kind == "c" || kind == "u") toml_skip_rest = 1
+        print line
+        return
+      }
+      encoded = line
+      gsub(/./, "&" char_separator, encoded)
+      split(encoded, chars, char_separator)
+      quote = ""
+      escaped = 0
+      field_allowed = 1
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:958]: Implemented in signed commit 1f1be11. Full-file Write/scan-path keep stateful checksum filtering; context-free Cargo/uv staged and apply_patch records remain fully scannable, and every Edit/MultiEdit/NotebookEdit fragment is scanned raw. Complete staged/apply_patch physical lines retain JSON/Yarn/go.sum grammar. Deterministic regressions cover Cargo staged/Edit/MultiEdit/apply_patch plus uv.lock and go.sum Edit custom-rule findings. The bundled default rules do not independently flag the exact nested sample, so the PR body documents this as a supported custom-rule preprocessor boundary.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

File: plugins/agent-guard/bin/agent-guard [hunk 3]
```diff
@@ -817,17 +817,230 @@ run_gitleaks_stdin() {
...
+        if (ch == "\"" || (kind != "p" && ch == single_quote)) {
+          quote = ch
+          escaped = 0
+          field_allowed = 0
+          i++
+          continue
+        }
+
+        if (ch == "{" || ch == ",")
+          field_allowed = 1
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Addressed in signed, GitHub-verified commit 0772ed0. Cargo neutralization is now limited to [[package]], legacy [root], and [[patch.unused]]; uv is limited to URL-bearing sdist/wheels artifacts under [[package]] or [[distribution]], including multi-line wheels; package-lock integrity is limited to direct package entries under packages/recursive dependencies. Context-free Cargo/uv/package-lock fragments remain raw, and JSON depth is capped fail-conservative. Custom-rule Write/scan-path/untracked/staged/apply_patch regressions plus direct schema probes pass; make test: 584/0.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

**Comment by JeongJaeSoon** [comment]
CodeRabbit’s latest review completed on 0772ed0 with one low-value maintainability nitpick and no behavioral or security finding. I am intentionally leaving the two small inline awk `lockfile_kind` functions duplicated: hoisting awk source into an interpolated shell variable would add shell/awk quoting and injection surface to this security fix, while the mappings are short, identical, and covered by staged/apply_patch regressions. A standalone refactor would be more appropriate if this mapping changes again.

**Comment by JeongJaeSoon** [comment]
@codex review

Please review signed, GitHub-verified head 0772ed099d6e5267904078c76aadd0f852f6184c. Focus on schema-scoped Cargo/uv/package-lock checksum neutralization, context-free fragment fail-conservative behavior, parser depth/line bounds, security bypasses, and false positives.

File: plugins/agent-guard/bin/agent-guard [hunk 4]
```diff
@@ -817,25 +817,341 @@ run_gitleaks_stdin() {
...
+                || (kind == "c" && ch == "c") \
+                || (kind == "u" && ch == "h")) \
+            && (kind != "p" || json_role[json_depth] == "entry") \
+            && (kind != "c" || cargo_checksum_scope) \
+            && (kind != "u" \
+                || (uv_artifact_line && brace_depth == 1 \
+                    && uv_url_seen[brace_depth])) \
+            && field_replacement(line, i, kind)) {
+          replacements[i] = replacement_text
+          skips[i] = replacement_length
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:1029]: Addressed in signed, GitHub-verified commit a059956. package-lock input is copied once into a private immutable snapshot; jq --stream validates exactly one complete top-level JSON object, and malformed, truncated, multi-document, or jq-unavailable input remains raw. Validation and filtering use the same snapshot to prevent atomic path-swap races, while snapshot-copy failure also falls back to raw scanning instead of empty stdin. Regressions cover proposed Write, scan-path, untracked and direct filtering, deterministic path replacement, and forced copy failure. make test: 590 passed, 0 failed; independent security review found no remaining actionable issue.

**Comment by JeongJaeSoon** [comment]
@coderabbitai review

@codex review

**Comment by JeongJaeSoon** [comment]
Independent security review found a binary lockfile fail-open path on the previous head: macOS awk truncated NUL-bearing records and could abort on invalid UTF-8, while child filter failures were not reliably propagated.

Addressed in signed commit `4fe569c` after merging current `main`:
- filter every supported lockfile from one immutable snapshot
- keep NUL-bearing input byte-for-byte raw for downstream scanning
- run record parsing in `LC_ALL=C` and preserve invalid UTF-8 plus the credential suffix
- fall back to the full snapshot if parsing fails
- propagate scan-path and untracked preparation failures
- add direct, real-gitleaks scan-path, and untracked regressions for NUL/invalid UTF-8

Verification: `make test` 600/0, smoke test, plugin layout, sh/dash syntax, and `git diff --check` all pass. Existing review threads remain resolved.

@codex review
@coderabbitai review

File: plugins/agent-guard/bin/agent-guard [hunk 5]
```diff
@@ -817,49 +817,474 @@ run_gitleaks_stdin() {
...
+      return "$status"
+    fi
+  fi
+  filtered=$(mktemp_file) || {
+    cat -- "$source"
+    status=$?
+    rm -f "$snapshot"
+    return "$status"
+  }
+  if filter_lockfile_records "$kind" "$source" >"$filtered"; then
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:1293]: Addressed in signed commit e99a505. The byte-locale awk parser now returns status 2 when a Cargo/uv stream reaches EOF (or changes record kind) with an open TOML multiline string. filter_lockfile_hashes discards any partial neutralized temp output and emits the immutable raw snapshot, so the checksum remains scannable. Added byte-for-byte raw-fallback and scan-path custom-scanner regressions for both Cargo.lock and uv.lock. Independent subagent re-review found no blocker; make test passes 604/0, plus smoke, layout, sh/dash syntax, submission-check, and diff-check.

**Comment by JeongJaeSoon** [comment]
@codex review
@coderabbitai review

Please review latest signed head 55096a6, including the incomplete-TOML raw-fallback fix in e99a505.

File: plugins/agent-guard/bin/agent-guard [hunk 6]
```diff
@@ -817,49 +817,474 @@ run_gitleaks_stdin() {
...
+      }
+      if (kind == "g") {
+        # Only the third go.sum field is a checksum. A module path may itself
+        # contain an h1-shaped secret and must remain visible to gitleaks.
+        if (match(line, /^[^[:space:]]+[[:space:]]+[^[:space:]]+[[:space:]]+/)) {
+          go_prefix = substr(line, 1, RLENGTH)
+          go_tail = substr(line, RLENGTH + 1)
+          if (go_tail ~ /^h1:[A-Za-z0-9+\/]{43}=([[:space:]]|$)/ \
+              && match(go_tail, /^h1:[A-Za-z0-9+\/]{43}=/))
+            line = go_prefix replace_match(go_tail, "h1:CHECKSUM")
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Addressed in signed commit 18dcbe0. go.sum neutralization now requires the complete remainder after the third-field h1 checksum to contain only record-ending whitespace; any trailing data remains byte-for-byte raw. Added direct raw-preservation and scan-path custom-scanner regressions. Independent subagent review found no blocker; full suite passes 610/0.

File: plugins/agent-guard/bin/agent-guard [hunk 7]
```diff
@@ -817,49 +817,474 @@ run_gitleaks_stdin() {
...
+  LC_ALL=C awk -v forced_kind="$forced_kind" '
+    function replace_match(line, replacement) {
+      return substr(line, 1, RSTART - 1) replacement substr(line, RSTART + RLENGTH)
+    }
+    function structured_delimiter(line, at, kind, n, ch) {
+      n = length(line)
+      if (kind == "y")
+        return at > n || substr(line, at, 1) ~ /[[:space:]]/
+      while (at <= n && substr(line, at, 1) ~ /[[:space:]]/) at++
+      return at > n || substr(line, at, 1) == "," || substr(line, at, 1) == "}"
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Addressed in signed commit 18dcbe0. Yarn integrity and Cargo checksum neutralization now validate the complete scalar-field remainder, accepting only end-of-record whitespace or an optional comment. Commas, braces, or other trailing data leave the full malformed record raw. Added byte-exact direct and scan-path regressions for both formats. Independent subagent review found no blocker; full suite passes 610/0.

**Comment by JeongJaeSoon** [comment]
@codex review
@coderabbitai review

Please review latest signed head 8c86987, including the malformed checksum-tail fixes in 18dcbe0.

**Comment by JeongJaeSoon** [comment]
Addressed CodeRabbit’s remaining incomplete-TOML P1 in signed commit `f49d6e4`:
- ordinary basic/literal strings and inline tables must close within their physical record
- multiline strings and arrays retain state through EOF
- unmatched delimiters and oversized/invalid TOML records force parser failure
- all failures discard partial neutralized output and scan the immutable raw snapshot
- Cargo/uv regressions cover multiline strings, ordinary quotes, inline tables, and arrays with byte-exact fallback plus scan-path detection

Independent subagent re-review found no blocker. Verification: `make test` 622/0, smoke, layout, sh/dash syntax, submission-check, and diff-check pass.

@codex review
@coderabbitai review

Please review latest signed head `7d737db`.

File: plugins/agent-guard/bin/agent-guard [hunk 8]
```diff
@@ -817,49 +817,490 @@ run_gitleaks_stdin() {
...
+    }
+    function structured_delimiter(line, at, kind, n, ch) {
+      n = length(line)
+      # Yarn integrity and Cargo checksum are whole scalar fields. Accept only
+      # end-of-record whitespace or a comment; malformed trailing data must
+      # remain raw and scannable. package-lock and uv use inline containers.
+      if (kind == "y" || kind == "c")
+        return substr(line, at) ~ /^[[:space:]]*(#.*)?$/
+      while (at <= n && substr(line, at, 1) ~ /[[:space:]]/) at++
+      return at > n || substr(line, at, 1) == "," || substr(line, at, 1) == "}"
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:841]: Addressed in signed commit f49d6e4. uv inline-table brace depth is now checked at every physical-record boundary; a hash line that ends without `}` sets invalid TOML state. EOF returns nonzero, partial filtered output is discarded, and the immutable raw snapshot is scanned. Cargo/uv inline-table regressions verify byte-exact fallback and scan-path detection. Independent subagent and final CodeRabbit reviews found no blocker; make test passes 622/0.

File: plugins/agent-guard/bin/agent-guard [hunk 9]
```diff
@@ -817,49 +817,490 @@ run_gitleaks_stdin() {
...
+        next
+      }
       print line
     }
-  '
+    END {
+      # Never emit checksum-neutralized output for incomplete TOML. The caller
+      # treats this parser failure as a signal to scan the full raw snapshot.
+      if ((previous_kind == "c" || previous_kind == "u") \
+          && toml_multiline != "")
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard]: Addressed in signed commit f49d6e4. Ordinary TOML quote state is now checked at each physical-record boundary; unterminated basic/literal strings set invalid state and force immutable raw-snapshot fallback at EOF. Cargo/uv ordinary-string regressions verify byte-exact fallback and scan-path detection. Independent subagent and final CodeRabbit reviews found no blocker; make test passes 622/0.

File: plugins/agent-guard/bin/agent-guard [hunk 10]
```diff
@@ -817,49 +817,517 @@ run_gitleaks_stdin() {
...
+          }
+          i++
+          continue
+        }
+
+        if (quote != "") {
+          if (quote == "\"" && ch == "\\") {
+            escaped = !escaped
+            i++
+            continue
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:996]: Addressed post-merge in follow-up PR #160, merged as f6b20851a8a5ccfb7966492a311be6f3cf82c4da. Invalid TOML basic-string escapes now force the immutable raw snapshot fallback instead of emitting checksum-neutralized content. The follow-up also covers CRLF multiline folds and truncated Unicode escapes. Verification: system awk 722/722, GNU awk 722/722, Ubuntu/macOS CI, Plugin Layout, HOL Scanner, lint, and submission validation all passed.
